### PR TITLE
[ML-33605] enabled cuda 11.8 for DCS image

### PR DESCRIPTION
**Pull request by ygong1**
Enabled cuda 11.8 for DCS Image. Since DBR 9.x, conda is no longer supported. Use virtualenv to build the cuda dependency as well as the dependency need to start databricks REPL. 

Tested
* pytorch - https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/756910245556995/command/756910245557007
* tensorflow - https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/756910245557420/command/756910245557421

### CODE REVIEW

File: ubuntu/gpu/README.md [hunk 1]
```diff
@@ -1,27 +1,23 @@
 # GPU Containers
 
-There are two variations of GPU containers that can be used depending upon the CUDA version you wish to use:
-[`cuda-10.1`](cuda-10.1) contains the layers which install CUDA 10.1
-[`cuda-11.0`](cuda-11.0) contains the layers which install CUDA 11.0
-
+**WARNING**: Using conda in DCS images is no longer supported since the Databricks Runtime 9.x. We highly recommend users to extend [`cuda-11.8`](cuda-11.8) examples.
+ We no longer support [`cuda-10.1`](cuda-10.1) and [`cuda-11.0`](cuda-11.0) complatibility with latest databricks runtime.
```
Comment (panchalhp-db) [inline ubuntu/gpu/README.md:4]:
```suggestion
 We no longer support [`cuda-10.1`](cuda-10.1) and [`cuda-11.0`](cuda-11.0) compatibility with latest databricks runtime.
```

File: ubuntu/gpu/README.md [hunk 2]
```diff
@@ -1,27 +1,23 @@
...
 Example containers for common GPU use cases:
-* [`gpu-tensorflow`](cuda-11.0/tensorflow) extends `gpu-conda` by creating a conda environment that contains [TensorFlow](https://www.tensorflow.org/).
-* [`gpu-pytorch`](cuda-11.0/pytorch) extends `gpu-conda` by creating a conda environment that contains [PyTorch](https://pytorch.org/).
-* [`gpu-rapids`](cuda-11.0/rapids) extends `gpu-conda` by creating a conda environment that contains [RAPIDS](https://rapids.ai/).
-  * RAPIDS requires NVIDIA Pascal GPU or better.
-    If you receive a `cudaErrorNoKernelImageForDevice: no kernel image is available for execution on the device` error,
-    you likely are using GPUs that are incompatible, e.g., K80 on EC2 P2 instances.
-    You should try switching to newer instance types.
-  * The Dockerfile provides an example to create the root conda environment from an environment spec file, which does not require dependency resolution.
+* [`gpu-tensorflow`](cuda-11.0/tensorflow) extends `gpu-venv` by creating a conda environment that contains [TensorFlow](https://www.tensorflow.org/).
```
Comment (panchalhp-db) [inline ubuntu/gpu/README.md:17]:
```suggestion
* [`gpu-tensorflow`](cuda-11.8/tensorflow) extends `gpu-venv` by creating a conda environment that contains [TensorFlow](https://www.tensorflow.org/).
```

File: ubuntu/gpu/cuda-11.8/venv/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  && rm get-pip.py
+
+
+# virtualenv 20.0.24 introduced a periodic update feature, which attempts to update all
+# seeder packages every 14 days. This launches background processes that may interfere
+# with user cleanup and may allow users to inadvertently update pip to newer versions
+# incompatible with Databricks. Instead, we patch virtualenv to disable periodic updates per
+# https://virtualenv.pypa.io/en/latest/user_guide.html#embed-wheels-for-distributions. See
+# ML-13720 for more details.
+RUN /usr/local/bin/pip3.10 install --no-cache-dir virtualenv==20.16.3 \
```
Comment (panchalhp-db) [inline ubuntu/gpu/cuda-11.8/venv/Dockerfile:21]: Should we use 20.16.7 to be consistent with DBR 14.0? https://github.com/databricks/universe/blob/master/spark/images/base/virtualenv-python/14.x/dbr-virtualenv-python.dockerfile.py#L22

File: ubuntu/gpu/cuda-11.8/venv/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+
+# These python libraries are used by Databricks notebooks and the Python REPL
+# You do not need to install pyspark - it is injected when the cluster is launched
+# Versions are intended to reflect latest DBR: https://docs.databricks.com/release-notes/runtime/11.1.html#system-environment
+RUN /databricks/python3/bin/pip install \
+  six==1.16.0 \
+  jedi==0.18.1 \
+  # ensure minimum ipython version for Python autocomplete with jedi 0.17.x
+  ipython==8.10.0 \
```
Comment (panchalhp-db) [inline ubuntu/gpu/cuda-11.8/venv/Dockerfile:132]: Shall we keep the versions the same as the ones we have in 14.0?

**Review by panchalhp-db — CHANGES_REQUESTED** [review]

File: ubuntu/gpu/cuda-11.0/base/Dockerfile [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # See https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.0.3/ubuntu18.04-x86_64/runtime/cudnn8/Dockerfile
-FROM nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu18.04
+FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04
```
[the remark this answers is not in the document]
Comment (ygong1) [reply ubuntu/gpu/cuda-11.0/base/Dockerfile:2]: oh, good catch, this is leaky experiment I did, will revert it back

File: ubuntu/gpu/cuda-11.8/build.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -e
+
+docker build -t databricksruntime/gpu-base:cuda11.8 base/
+docker build -t databricksruntime/gpu-conda:cuda11.8 conda/
+docker build -t databricksruntime/gpu-tensorflow:cuda11 tensorflow/
```
[the remark this answers is not in the document]
Comment (ygong1) [reply ubuntu/gpu/cuda-11.8/build.sh:7]: good catch

File: ubuntu/gpu/README.md [hunk 7]
```diff
@@ -1,37 +1,33 @@
 # GPU Containers
 
-There are two variations of GPU containers that can be used depending upon the CUDA version you wish to use:
-[`cuda-10.1`](cuda-10.1) contains the layers which install CUDA 10.1
-[`cuda-11.0`](cuda-11.0) contains the layers which install CUDA 11.0
+**WARNING**: Using conda in DCS images is no longer supported starting Databricks Runtime 9.0. We highly recommend users to extend [`cuda-11.8`](cuda-11.8) examples.
+ We no longer support [`cuda-10.1`](cuda-10.1) and [`cuda-11.0`](cuda-11.0) compatibility with latest databricks runtime.
 
+**WARNING**: DCS images which extends [`venv`](cuda-11.8/venv/) need to be paired with DBR 14.x when creating a
```
Comment (panchalhp-db) [inline ubuntu/gpu/README.md:6]:
```suggestion
```suggestion
**WARNING**: DCS images which extends [`venv`](cuda-11.8/venv/) need to be paired with DBR 14.x when creating a
```
**WARNING**: DCS images which extends [`venv`](cuda-11.8/venv/) need to be paired with DBR 14.0 and above when creating a
```

File: ubuntu/gpu/README.md [hunk 8]
```diff
@@ -1,37 +1,33 @@
...
+cluster. In order for REPL to launch successfully, the python packages used in the image HAVE to match with python packges used in the paried DBR version. [`venv`](cuda-11.8/venv/) uses version set from DBR 14.x
```
Comment (panchalhp-db) [inline ubuntu/gpu/README.md:7]:
```suggestion
cluster. In order for REPL to launch successfully, the python packages used in the image HAVE to match with python packges used in the paried DBR version. [`venv`](cuda-11.8/venv/) uses version set from DBR 14.0
```

File: ubuntu/gpu/cuda-11.8/venv/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  Jinja2==3.1.2\
+  ipykernel==6.25.0 \
+  protobuf==4.23.3 \
+  grpcio==1.48.2 \
+  grpcio-status==1.48.1 \
+  databricks-sdk==0.1.6
+
+
+
+# Specifies where Spark will look for the python process
```
Comment (panchalhp-db) [inline ubuntu/gpu/cuda-11.8/venv/Dockerfile:53]:
```suggestion
# Specifies where Spark will look for the python binary
```

File: ubuntu/gpu/cuda-11.8/venv/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+# Create /databricks/python3 environment.
+# We install pip and wheel so their executables show up under /databricks/python3/bin.
+# We use `--system-site-packages` so python will fallback to system site packages.
+# We use `--no-download` so virtualenv will install the bundled pip and wheel.
+RUN virtualenv --python=/usr/bin/python3.10 /databricks/python3 --system-site-packages --no-download
+
+
+# These python libraries are used by Databricks notebooks and the Python REPL
+# You do not need to install pyspark - it is injected when the cluster is launched
+# Versions are intended to reflect DBR 14.x: https://docs.databricks.com/release-notes/runtime/releases.html
```
Comment (panchalhp-db) [inline ubuntu/gpu/cuda-11.8/venv/Dockerfile:34]:
```suggestion
# Versions are intended to reflect DBR 14.0: https://docs.databricks.com/release-notes/runtime/releases.html
```

**Review by panchalhp-db — APPROVED** [review]
LGTM. Just a few minor comments.

**Pull request merged**